akka-sample-distributed-workers-scala/src/main/scala/worker/WorkManager.scala

```diff
@@ -19,169 +19,124 @@ import worker.WorkState.WorkerFailed
 
-  sealed trait WorkerStatus
-  case object Idle extends WorkerStatus
-  final case class Busy(workId: String, deadline: Deadline) extends WorkerStatus
-  final case class WorkerState(ref: ActorRef[WorkerCommand], status: WorkerStatus)
+//  sealed trait WorkerStatus
+//  case object Idle extends WorkerStatus
+//  final case class Busy(workId: String, deadline: Deadline) extends WorkerStatus
+//  final case class WorkerState(ref: ActorRef[WorkerCommand], status: WorkerStatus)
 
-  sealed trait Command extends CborSerializable
-  private final case class UpdatedWorkers(workers: Receptionist.Listing) extends Command
 
   // Messages from Workers
-  final case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[WorkerCommand]) extends Command
-  final case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[WorkerCommand])
-      extends Command
-  final case class WorkFailed(worker: ActorRef[WorkerCommand], workId: String) extends Command
+ // final case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[WorkerCommand]) extends Command
+  //final case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[WorkerCommand])
+  //    extends Command
+  //final case class WorkFailed(worker: ActorRef[WorkerCommand], workId: String) extends Command
 
   // Responses to requests from workers
   sealed trait WorkerCommand
   final case class DoWork(work: Work) extends WorkerCommand
-  final case class WorkAck(id: String) extends WorkerCommand
-  case object WorkAvailable extends WorkerCommand
 
-  // External commands
+  sealed trait Command extends CborSerializable
   final case class SubmitWork(work: Work, replyTo: ActorRef[WorkManager.Ack]) extends Command
+  private case class RequestNextWrapper(ask: RequestNext[WorkerCommand]) extends Command
+  final case class WorkIsDone(id: String) extends Command
+  final case class WorkFailed(id: String, t: Throwable) extends Command
+  private final case class TryStartWork() extends Command
 
   def apply(workTimeout: FiniteDuration): Behavior[Command] =
     Behaviors.setup { ctx =>
       Behaviors.withTimers { timers =>
         // FIXME use typed pub sub once https://github.com/akka/akka/issues/26338 is done
+
+        implicit val timeout = Timeout(5.seconds)
         val mediator = DistributedPubSub(ctx.system.toClassic).mediator
+        val producerController = ctx.spawn(WorkPullingProducerController[WorkerCommand]("work-manager", ManagerServiceKey, None), "producer-controller")
+        val requestNextAdapter = ctx.messageAdapter(RequestNextWrapper)
+        producerController ! WorkPullingProducerController.Start(requestNextAdapter)
 
-        // the set of available workers is not event sourced as it depends on the current set of workers
-        var workers = Map[ActorRef[WorkerCommand], WorkerState]()
+        var requestNext = Queue[RequestNext[WorkerCommand]]()
```

Comment:
not sure I understand why this would have to be a Queue? isn't it enough with Option?

Reply:
That was my misunderstanding of how the fan out producer worked